File: project/app/src/main/java/org/owntracks/android/services/MessageProcessorEndpoint.java
package org.owntracks.android.services;

import org.owntracks.android.messages.MessageBase;
import org.owntracks.android.support.interfaces.ConfigurationIncompleteException;
import org.owntracks.android.support.interfaces.OutgoingMessageProcessor;

import java.io.IOException;

public abstract class MessageProcessorEndpoint implements OutgoingMessageProcessor {
    MessageProcessor messageProcessor;

    MessageProcessorEndpoint(MessageProcessor messageProcessor) {
        this.messageProcessor = messageProcessor;
    }

    void onMessageReceived(MessageBase message) {
        message.setIncoming();
        message.setModeId(getModeId());
        messageProcessor.processIncomingMessage(onFinalizeMessage(message));
    }

    protected abstract MessageBase onFinalizeMessage(MessageBase message);

    abstract int getModeId();

    abstract void sendMessage(MessageBase m) throws ConfigurationIncompleteException, OutgoingMessageSendingException, IOException;
}

class OutgoingMessageSendingException extends Exception {
    OutgoingMessageSendingException(Exception e) {
        super(e);
    }
}

File: project/app/src/main/java/org/owntracks/android/services/MessageProcessorEndpoint.kt
package org.owntracks.android.services

import java.io.IOException
import org.owntracks.android.model.messages.MessageBase
import org.owntracks.android.preferences.types.ConnectionMode
import org.owntracks.android.support.interfaces.ConfigurationIncompleteException
import org.owntracks.android.support.interfaces.OutgoingMessageProcessor

abstract class MessageProcessorEndpoint internal constructor(val messageProcessor: MessageProcessor) :
    OutgoingMessageProcessor {
    suspend fun onMessageReceived(message: MessageBase) {
        message.setIncoming()
        message.modeId = modeId!!
        messageProcessor.processIncomingMessage(onFinalizeMessage(message))
    }

    protected abstract fun onFinalizeMessage(message: MessageBase): MessageBase
    abstract val modeId: ConnectionMode?

    @Throws(
        ConfigurationIncompleteException::class,
        OutgoingMessageSendingException::class,
        IOException::class
    )
    abstract suspend fun sendMessage(message: MessageBase)
}

class OutgoingMessageSendingException internal constructor(e: Exception?) : Exception(e)
